fastestimator/schedule/schedule.py

```diff
@@ -205,7 +210,31 @@ def get_current_items(items: Iterable[Union[T, Scheduler[T]]],
         else:
             item = [item]
         for item_ in item:
-            if item_ and (not run_modes or not hasattr(item_, "mode") or not item_.mode
-                          or item_.mode.intersection(run_modes)):
+            # mode matching
+            mode_match = False
+            if not run_modes:
+                mode_match = True
+            if not hasattr(item_, "mode"):
+                mode_match = True
+            else:
+                if not item_.mode:
+                    mode_match = True
+                elif item_.mode.intersection(run_modes):
+                    mode_match = True
+
+            # ds_id matching
+            ds_id_match = False
+            if not ds_id:
+                ds_id_match = True
+            if not hasattr(item_, "ds_id"):
+                ds_id_match = True
+            else:
+                if not item_.ds_id:
+                    ds_id_match = True
+                elif ds_id in item_.ds_id:
+                    ds_id_match = True  # whitelist check
+                elif any([x.startswith("!") for x in item_.ds_id]) and all([ds_id != x[1:] for x in item_.ds_id]):
```

Comment:
might want inline comment about how if any ds_id is negated then they all must be negated. This is another place where if we could auto-resolve them to the positive set we could simplify

Reply:
added comment